### DISCUSSION TITLE
Async Autocomplete is empty onChange

### DISCUSSION THREAD

**Discussion by mergepls** [mantinedev/mantine#822] [Q&A]
I am trying to use the `Autocomplete` component with data from an API using `react-query`. I am doing a server-side search, however it seems like the Autocomplete component is not playing well with async data.

Using `console.log(data?.Ingredients.map(buildOptions)`, I do see data returned, but the component keeps displaying the `nothingFound` component.

```
import React, { forwardRef } from "react";
import { Autocomplete, Group, Loader, Text } from "@mantine/core";
import { useDebouncedValue } from "@mantine/hooks";
import { useEffect, useState } from "react";
import { client } from "../../graphql/client";
import { Ingredient, useGetIngredientsQuery } from "../../graphql/generated";

interface IngredientItem extends Ingredient {
  value: string;
}

const IngredientsAutocompleteItem = forwardRef<HTMLDivElement, IngredientItem>(
  ({ category, name, value, ...others }: IngredientItem, ref) => (
    <div ref={ref} {...others}>
      <Group noWrap>
        <div>
          <Text>{name}</Text>
          <Text size="xs" color="dimmed">
            {category.name}
          </Text>
        </div>
      </Group>
    </div>
  )
);

function IngredientsAutocomplete() {
  const [value, setValue] = useState("");
  const [result, setResult] = useState<IngredientItem[]>([]);
  const [debouncedValue] = useDebouncedValue(value, 250);

  const { data } = useGetIngredientsQuery(client, {
    searchString: debouncedValue,
  });
  
  function buildOptions(ingredient: Ingredient) {
    return {
      ...ingredient,
      value: ingredient.id,
    };
  }

  return (
    <Autocomplete
      nothingFound={<Text>Not found</Text>}
      itemComponent={IngredientsAutocompleteItem}
      placeholder="Search ingredients to add"
      value={value}
      onChange={setValue}
      data={data?.Ingredients.map(buildOptions) || []}
    />
  );
}

export { IngredientsAutocomplete };
```

**Comment by rtivital** [comment]
Make sure your data fetching strategy triggers component rerender, here is dynamic data example – https://codesandbox.io/s/youthful-bird-7djso?file=/src/App.tsx

**Reply by mergepls** [reply]
Found the culprit:

https://github.com/mantinedev/mantine/blob/0ac9e11b67ad3842ead131b75287f439b6e34527/src/mantine-core/src/components/Autocomplete/Autocomplete.tsx#L34

In my case, Autocomplete is doing a client side filter regardless if you pass the `filter` prop. It uses the `value` field of the AutocompleteItem and since I have my `value` is set to `value: ingredient.id` it was not matching anything the text in the input (name).

In my case, I am doing a server side filter so there is no need for the additional client side filter. I did solve this by passing `filter={(value, item) => true}` as props. I suggest introducing a props to disable the client-side filtering.

**Reply by rtivital** [reply]
`filter` prop is made for that case, not need to introduce other props

**Reply by ericschmar** [reply]
@mergepls @rtivital I ran into this same exact issue. Took all morning to figure this out. @mergepls has the winning code for me. I don't think he's suggesting adding a new prop, but you could make `filter` optional. As it stands, there's no indication in the docs that it defaults the the `.id` of the item. So either remove the hidden functionality, or update the docs to tell people that is does in fact filter your list even without explictly having it do that.

**Comment by Fedec96** [comment]
Just ran into this issue myself. As of Mantine v6, `filter={() => true}` is still the only solution for when filtering is performed server-side. Any chance this is changing in the future?

**Reply by rtivital** [reply]
Why this needs to be changed?

**Reply by mergepls** [reply]
Nothing needs to be changed except for maybe the docs for this component.

When performing a server-side filtering, we don't need the additional client-side filtering provided by the Autocomplete component. Hence, why I suggested to disable client-side filtering using `filter={() => true}`.

IMO, it is worth mentioning the default filtering behavior in the docs so that have a better understanding of how the Autocomplete component behaves. Happy to take a stab at updating the docs.

**Reply by kashpatel** [reply]
Ran into the same issue. Would be nice if docs reflected this. Thank you.

**Reply by 0xFFrancesco** [reply]
+1! 👍

Had the same exact issue, and couldn't understand why it wasn't working before seeing this.

I also use a server-side filtering via react-query. The data was flowing right, but the component didn't display it.

It was for the same reason above: I was using a custom `itemComponent`, and the data `value` was set to an ID, so it was not matched by the input `value`.

@rtivital 

I think it should be made more clear in the docs. Maybe just a little paragraph saying that: _by default the `data` prop (and especially the `value` field) is filtered by the input `value`, and should probably be disabled in case the data is already filtered outside of the component itself_. Maybe also add a prop like `disableAutoFiltering` instead of `filter={() => true}` seems cleaner.

**Reply by engin-can** [reply]
I'm getting `Type '() => boolean' is not assignable to type 'OptionsFilter'.
  Type 'boolean' is not assignable to type 'ComboboxParsedItem[]'.ts(2322)` on `filter={() => true}` with mantine v7.11.1

**Reply by rtivital** [reply]
https://mantine.dev/core/autocomplete/#options-filtering
